lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt

```diff
@@ -1116,64 +1109,113 @@ internal class EvaluatingCompiler(
-                        // check if FLOAT has been given an argument--throw exception since we do not honor it.
-                        if (expr.type.args.any()) {
-                            err(
-                                "FLOAT precision parameter is unsupported",
-                                ErrorCode.SEMANTIC_FLOAT_PRECISION_UNSUPPORTED,
-                                errorContextFrom(expr.type.metas),
-                                internal = false)
+        return when (compileOptions.typedOpBehavior) {
+            TypedOpBehavior.LEGACY -> simpleTypeMatchFunc
+            TypedOpBehavior.HONOR_PARAMETERS -> { expValue: ExprValue ->
+                staticType.allTypes.any {
+                    val matchesStaticType = try {
+                        it.isInstance(expValue)
+                    } catch (e: UnsupportedTypeCheckException) {
+                        err(
+                            e.message!!,
+                            ErrorCode.UNIMPLEMENTED_FEATURE,
+                            errorContextFrom(metas),
+                            internal = true
+                        )
+                    }
+
+                    when {
+                        !matchesStaticType -> false
+                        else -> when (val validator = typedOpParameter.validationThunk) {
+                            null -> true
+                            else -> validator(expValue)
                         }
-                    else -> {
-                        // no validation needed since we honor all other relevant type parameters.
                     }
                 }
             }
         }
+    }
 
-        when (typedOpParameter.staticType) {
-            is AnyType -> {
-                // return trivial results for operations against ANY
-                return when (op) {
-                    TypedOp.CAST -> expThunk
-                    TypedOp.CAN_CAST, TypedOp.CAN_LOSSLESS_CAST, TypedOp.IS ->
-                        thunkFactory.thunkEnv(metas) { valueFactory.newBoolean(true) }
+    private fun compileIs(expr: PartiqlAst.Expr.IsType, metas: MetaContainer): ThunkEnv {
+        val expThunk = compileAstExpr(expr.value)
+        val typedOpParameter = expr.type.toTypedOpParameter(customTypedOpParameters)
+        if (typedOpParameter.staticType is AnyType) {
+            return thunkFactory.thunkEnv(metas) { valueFactory.newBoolean(true) }
+        }
+        if (compileOptions.typedOpBehavior == TypedOpBehavior.HONOR_PARAMETERS && expr.type is PartiqlAst.Type.FloatType && expr.type.precision != null) {
+            err(
+                "FLOAT precision parameter is unsupported",
+                ErrorCode.SEMANTIC_FLOAT_PRECISION_UNSUPPORTED,
+                errorContextFrom(expr.type.metas.toPartiQlMetaContainer()),
+                internal = false)
+        }
+
+        val typeMatchFunc = when (val staticType = typedOpParameter.staticType) {
+            is AnyType -> { _: ExprValue -> true }
```

Comment:
This should be an `IllegalStateException` as it was previously: https://github.com/partiql/partiql-lang-kotlin/blob/7f07dc31570f4e10e4e433a8878e35416771d6f7/lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt#L1392-L1393

In the current PR, `typedOpParameter.staticType` is already checked earlier for being `AnyType` on L1164-1166:
https://github.com/partiql/partiql-lang-kotlin/blob/6ec9e0ad24b29152e94db19df073e981c652059b/lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt#L1164-L1166. So that branch of the `when` expression should result in an `IllegalStateException`.

Reply:
In the original code, the check for `AnyType` already happens in [L1127](https://github.com/partiql/partiql-lang-kotlin/blob/7f07dc31570f4e10e4e433a8878e35416771d6f7/lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt#L1127). And this is the reason [L1392](https://github.com/partiql/partiql-lang-kotlin/blob/7f07dc31570f4e10e4e433a8878e35416771d6f7/lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt#L1392-L1393) saying "it should never happen". However, in the current PR, `compileIs` is directly called from `compileAstExpr`, (not from `compileTyped`), so the check for `AnyType` should behave the same as [L1127](https://github.com/partiql/partiql-lang-kotlin/blob/7f07dc31570f4e10e4e433a8878e35416771d6f7/lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt#L1127)

Reply:
(Using the latest commit, [b9ed8c9](https://github.com/partiql/partiql-lang-kotlin/commit/b9ed8c9e2c45dd9f8f56c8be98de21ddd77f1998)): In `compileIs`, `typedOpParameter.staticType` is checked to be `AnyType` on L1173-L1175 https://github.com/partiql/partiql-lang-kotlin/blob/b9ed8c9e2c45dd9f8f56c8be98de21ddd77f1998/lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt#L1173-L1175. If it's `AnyType`, then `compileIs` returns. So the subsequent `when` expression checking for `AnyType` will never run. That's why I'm suggesting it makes more sense to give the `IllegalStateException` since that line will never run.

Reply:
I got it. Will replace L1185 with `IllegalStateException`.